[User-Written Triton] Handle the `scf.for` and `scf.while` case

sijiac opened the pull request:
Summary:
This is the official fix of the issue, reported in https://fb.workplace.com/groups/1075192433118967/permalink/1427865377851669/

The root-cause is the MLIR mutation analyze doesn't find the mutated tensors, which made AOT autograd think there is no users of the Triton kernel and then removed it 😔

---

Triton IR: P1369315213
Wrong Analyze Graph: P1364305956
Right Analyze Graph: P1369324977

Test Plan:
buck2 run mode/opt scripts/liptds/domain_kernels:triton_dcpp_flash

unit tests

Differential Revision: D57606053




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

oulgen commented on torch/_higher_order_ops/triton_kernel_wrap.py:
```diff
@@ -288,7 +288,27 @@ def mlir_to_functions(op) -> None:
+                                next_fake_intermediate -= 1
+                                replacements[idx] = Intermediate(next_fake_intermediate)
+                            else:
+                                replacements[idx] = Intermediate(operand_ids[i+2])
```
can you explain why there's a +2 here?

sijiac replied:
operands[0] and operands[1] are lower-bound and upper-bound, where we don't have corresponding block args. The ^^ comments in code should give the example

chenyang78 replied:
I think it refers to ```%init``` when ```i == 1```. I would replace the code with something like below:

```
nxt_fake_intermediate -= 1
idx = block_id_to_block_arg_ids[block_id][0]
replacements[idx] = Intermediate(next_fake_intermediate)
idx = block_id_to_block_arg_ids[block_id][1]
replacements[idx] = Intermediate(operand_ids[3])
```

sijiac replied:
ohh, the num block args are not guaranteed to 2, it should be >= 1 based on the behavior

chenyang78 commented:
Also, please fix the lint issue :) Thanks!

sijiac committed:
[User-Written Triton] Handle the `scf.for` and `scf.while` case (#127065)

Summary:

This is the official fix of the issue, reported in https://fb.workplace.com/groups/1075192433118967/permalink/1427865377851669/

The root-cause is the MLIR mutation analyze doesn't find the mutated tensors, which made AOT autograd think there is no users of the Triton kernel and then removed it 😔

---

Triton IR: P1369315213
Wrong Analyze Graph: P1364305956
Right Analyze Graph: P1369324977

Test Plan:
buck2 run mode/opt scripts/liptds/domain_kernels:triton_dcpp_flash

unit tests

Reviewed By: oulgen

Differential Revision: D57606053

sijiac commented:
@pytorchbot label  release notes:fix the issue of user-written triton kernels are silently removed

sijiac commented:
@pytorchbot merge

sijiac commented:
@pytorchbot label "release notes:autograd"

sijiac commented:
@pytorchbot label "release notes: autograd"

sijiac commented:
@pytorchbot merge
